File: python/tree/leetocde/symmetric_tree.py

# 对称二叉树
# https://leetcode-cn.com/problems/symmetric-tree/


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    # 递归的方式，深度优先
    def isSymmetric_1(self, root: TreeNode) -> bool:
        def helper(a, b) -> bool:
            if not a and not b:
                return True
            if not a or not b or a.val != b.val:
                return False
            return helper(a.left, b.right) and helper(a.right, b.left)
        return helper(root, root)

    # 使用一个队列，广度优先
    def isSymmetric_2(self, root: TreeNode) -> bool:
        def helper(node1: TreeNode, node2: TreeNode) -> bool:
            queue = [node1, node2]
            while len(queue) > 0:
                n1, n2 = queue.pop(0), queue.pop(0)
                if not n1 and not n2:
                    continue
                if not n1 or not n2 or n1.val != n2.val:
                    return False
                queue.append(n1.left), queue.append(n2.right)
                queue.append(n1.right), queue.append(n2.left)
            return True
        return helper(root, root)

File: go/tree/leetcode/symmetric_tree.go
package leetcode

//递归
func isSymmetric1(root *TreeNode) bool {
	return symmetricHelper(root, root)
}

func symmetricHelper(left, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil || left.Val != right.Val {
		return false
	}

	return symmetricHelper(left.Left, right.Right) && symmetricHelper(left.Right, right.Left)
}

//广度优先
func isSymmetric2(root *TreeNode) bool {
	queue := []*TreeNode{root, root}

	for len(queue) > 0 {
		x, y := queue[0], queue[1]
		queue = queue[2:]
		if x == nil && y == nil {
			continue
		}
		if x == nil || y == nil || x.Val != y.Val {
			return false
		}

		queue = append(queue, x.Left, y.Right, x.Right, y.Left)
	}

	return true
}
